compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributionPatterns.cpp

```diff
@@ -369,9 +370,161 @@ struct DistributeTransferWriteLayoutAttr final
+
+    // Reduce packed vector with initial value.
+    Value reducedValue = rewriter.create<vector::ExtractOp>(
+        loc, result, SmallVector<int64_t>{0});
+    for (int i = 1; i < entriesPerVector; i++) {
+      Value next = rewriter.create<vector::ExtractOp>(loc, result,
+                                                      SmallVector<int64_t>{i});
+      reducedValue = makeArithReduction(rewriter, loc, combiningKind,
+                                        reducedValue, next, nullptr, mask);
+    }
+    result = makeArithReduction(rewriter, loc, combiningKind, reducedValue,
+                                mEmpty, nullptr, mask);
+    return result;
+  }
+
+  // This pattern distributes reductions as follows:
+  // First, the data local to a specific thread is reduced.
+  // Then, the data between threads is reduced by emitting appropriate
+  // shuffle instructions.
+  // Currently, only 16 and 32 bit types are supported.
+  // TODO: Add ability to reduce n parallel dims together.
+  LogicalResult matchAndRewrite(vector::MultiDimReductionOp reductionOp,
+                                DistributionSignature &signature,
+                                PatternRewriter &rewriter) const override {
+    auto reductionDims = llvm::to_vector<4>(
+        reductionOp.getReductionDims().getAsRange<IntegerAttr>());
+    // TODO: Add support for reductions along multiple dimensions.
+    if (reductionDims.size() > 1)
+      return failure();
+
+    VectorValue resultVec = dyn_cast<VectorValue>(reductionOp.getResult());
+    // TODO: Support results that are not vectors.
+    if (!resultVec)
+      return failure();
+    LayoutAttr resultLayout = dyn_cast<LayoutAttr>(signature[resultVec]);
+    if (!resultLayout)
+      return failure();
+
+    VectorValue source = reductionOp.getSource();
+    ShapedType sourceType = llvm::cast<ShapedType>(source.getType());
+    // TODO: Add support for (n != 2)-D tensors.
+    if (sourceType.getRank() != 2)
+      return failure();
+
+    LayoutAttr sourceLayout = dyn_cast<LayoutAttr>(signature[source]);
+    if (!sourceLayout)
+      return failure();
+
+    VectorValue acc = dyn_cast<VectorValue>(reductionOp.getAcc());
+    ShapedType accType = llvm::cast<ShapedType>(acc.getType());
+    Type elementType = accType.getElementType();
+    int bitWidth = elementType.getIntOrFloatBitWidth();
+    // TODO: Support additional bitwidths.
+    if ((bitWidth != 16) && (bitWidth != 32))
+      return failure();
+
+    Location loc = reductionOp.getLoc();
+    auto storeVectorType =
+        VectorType::get(resultLayout.getDistributedShape(), elementType);
+    Value storeVec = rewriter.create<arith::ConstantOp>(
+        loc, storeVectorType, rewriter.getZeroAttr(storeVectorType));
+
+    int reductionDim = reductionDims[0].getInt();
+    int parallelDim = reductionDim ^ 1;
+    uint64_t shuffleOffset = sourceLayout.getShuffleOffset(reductionDim);
+    if (!sourceLayout.getLane(reductionDim))
+      return failure();
```

Comment:
nit: getShuffleOffset internally looks for `reductionDim`
```suggestion
    if (!sourceLayout.getLane(reductionDim))
      return failure();
    uint64_t shuffleOffset = sourceLayout.getShuffleOffset(reductionDim);
```

Reply:
thanks will change